### DISCUSSION TITLE
How to resolve memory leak when integrating Langfuse with LangChain

### DISCUSSION THREAD

**Discussion by ivan-94** [langfuse/langfuse#3901] [Support] [answered]
I've encountered a memory leak issue when integrating Langfuse with LangChain. Here are some details:

```python
from langchain_core.language_models import FakeListChatModel
import gc
import time

model = FakeListChatModel(responses=["Hello, how can I assist you today?"])

for i in range(100):
    model.invoke(
        "Hello, how can I assist you today?" * 500,
        config={"callbacks": [LangfuseCallbackHandler()]},
    )

time.sleep(1)

gc.collect()
```

Memray report:
![WX20240820-110856@2x](https://github.com/user-attachments/assets/8160eea1-261d-4515-bb62-917eae5d69c9)

objgraph:
<img width="1210" alt="image" src="https://github.com/user-attachments/assets/479b535b-0625-496a-bd4c-3e368587dba6">

### To reproduce

Set up a basic LangChain application
Integrate Langfuse for tracing/logging
Run the application for an extended period
Monitor memory usage

### SDK and container versions

langfuse 2.43.3
langchain-core = "^0.2.33"
langchain = "^0.2.14"
python = "^3.11"

### Additional information

Langfuse reporting forks multiple consumer threads.
These threads remain in an active state and are not being released.
The accumulation of these unreleased threads leads to a memory leak over time.

**Comment by marcklingen** [comment, marked as answer]
Thanks for reporting, if you do not create a callback handler for each invoke, you will not get this memory leak:

```
from langchain_core.language_models import FakeListChatModel
import gc
import time
 
model = FakeListChatModel(responses=["Hello, how can I assist you today?"])

callback_handler = LangfuseCallbackHandler()

for i in range(100):
    model.invoke(
        "Hello, how can I assist you today?" * 500,
        config={"callbacks": [callback_handler]},
    )
 
time.sleep(1)
 
gc.collect()
```

Thereby, a single background process is used to capture traces and send them in batches to the langfuse api

**Comment by johschmidt42** [comment]
@marcklingen 

If there should be only a single callback_handler (only one instance in comparison to 1 handler per invocation), how should we update fields such as `user_id` or `session_id`? Wouldn't updating the IDs in a shared handler lead to race conditions and incorrect data if we have multiple requests that are processed in parallel (instead of sequentially)?

**Comment by marcklingen** [comment]
@johschmidt42 these parameters can now be set via the langchain metadata to make this race-condition/thread safe. Have a look here: https://langfuse.com/docs/integrations/langchain/tracing#dynamic-trace-attributes-in-chain-invocation

**Comment by johschmidt42** [comment]
Thanks @marcklingen, I just found this too. 

All that is required, is to add a config like this:

```python
handler = LangfuseCallbackHandler()

config = RunnableConfig(
        callbacks=[handler],
        metadata={
            "langfuse_session_id": operation_id,
            "langfuse_user_id": user_id,
        },
    )
    
chain.invoke("Some input", config=config)   
```

**Comment by marcklingen** [comment]
Great, did this solve the problem for you?

**Comment by johschmidt42** [comment]
Absolutely.

**Comment by willxie** [comment]
@marcklingen I am debugging a memory leak issue that led me to this thread. I was wondering what's the @observe compatible version of this fix and what the best practices are. My current code is like this:
```
default_langfuse_handler = CallbackHandler(
    secret_key=LANGFUSE_SECRET_KEY,
    public_key=LANGFUSE_PUBLIC_KEY,
    host=LANGFUSE_HOST,
)

def get_langfuse_callback_config():
    try:
        handler = langfuse_context.get_current_langchain_handler()
        if handler is not None:
            return {"callbacks": [handler]}
        else:
            return {"callbacks": [default_langfuse_handler]}
    except Exception as e:
        return {"callbacks": [default_langfuse_handler]}
```

I call it like so throughout my codebase, usually with `@observe` decorator far remove the LLM calls. 
```
await chain.ainvoke(
        ...
        config=get_langfuse_callback_config(),
    )
```
I run this on Cloud Run with FastAPI where I am experiencing a memory leak that pointed me to langfuse being the main culprit. So I was wondering what are the best ways to structure langfuse with langchain while making sure that the logs are flushed and memory deallocated.
